addon/Simulator.jsm

```diff
@@ -0,0 +1,142 @@
+ * @param {string} rapporPath - Path where the RAPPOR simulator lives.
+ * @param {object} params - Object containing the algorithm parameters.\
+ * @param {constant} method  - Method used to encode. Supported: nsICryptoHash.MD5 and nsICryptoHash.SHA256
+ * @param {string} instance - Simulation instance.
+ */
+function runRapporSimulation(studyName, rapporPath, params, method, instance) {
+  let data = read(new FileUtils.File(rapporPath + "_tmp/python/" + instance + "/1/case_true_values.csv"));
+  let caseReportsFile = new FileUtils.File(rapporPath + "_tmp/python/" + instance + "/1/case_reports.csv");
+  write(caseReportsFile, "client, cohort, bloom, prr, irr\n");
+  // Iterate over each line of the file getting the client, cohort and value.
+  // Each line represents a candidate sample that will be used to test the rappor implementation.
+  for (let i = 1; i < data.length; i++) {
+    let line = data[i].split(",");
+    let report = TelemetryRappor.createReport(studyName, line[2], params, method, line[1]);
+    // The expected format is {client, cohort, bloom, prr, irr}.
+    write(caseReportsFile, line[0] + ","+ line[1] + "," + convertToBin(report.internalBloom) +
+          "," + convertToBin(report.internalPrr) + "," + convertToBin(report.report) + "\n");
+  }
+}
+
+/**
+ * Read the cohort and the true value from a file. {client, cohort, value}.
+ * @param {nsFile} file - file containing the true values.
+ */
+function read(file) {
+  // open an input stream from file
+  var istream = Cc["@mozilla.org/network/file-input-stream;1"].createInstance(Ci.nsIFileInputStream);
+  istream.init(file, 0x01, 0o444, 0);
+  istream.QueryInterface(Ci.nsILineInputStream);
+  // read lines into array
+  var line = {}, lines = [], hasmore;
+  do {
+    hasmore = istream.readLine(line);
+    lines.push(line.value);
+  } while(hasmore);
+
+  istream.close();
+
+  return lines;
+}
+
+/**
+ * Write in a CSV {client, cohort, bloom, prr, irr}.
+ * @param {nsIFile} file - File to write in.
+ * @param {string} data - String containing the client, cohort, bloom, prr and irr.
+ */
+function write(file, data) {
+  // file is nsIFile, data is a string
+  var foStream = Cc["@mozilla.org/network/file-output-stream;1"].createInstance(Ci.nsIFileOutputStream);
+
+  foStream.init(file, 0x02 | 0x08 | 0x10, 0o666, 0); // write | create | append
+  var converter = Cc["@mozilla.org/intl/converter-output-stream;1"].createInstance(Ci.nsIConverterOutputStream);
+  converter.init(foStream, "UTF-8", 0, 0);
+  converter.writeString(data);
+  converter.close(); // this closes foStream
+}
+
+/**
+ * Convert an string representing hex into a binary string.
+ * @param hex - hex string to convert.
+ */
+function convertToBin(hex) {
+  let str = parseInt(hex, 16).toString(2);
+  let expected = hex.toString().length * 4;
+  let real = str.length;
+  while (real < expected) {
+    str = '0' + str;
+    real++;
+  }
+  return str;
+}
+
+var Simulator = {
+/**
```

Comment:
The comments and the function are not properly indented.

Reply:
Done